File: virtconsole.js
var Buffer = require('buffer').Buffer,
    fs = require('fs'),
    http = require('http'),
    net = require('net'),
    path = require('path'),
    url = require('url'),
    libvirt = require('/usr/lib/nodejs/libvirt/lib/libvirt.node'),
    WebSocketServer = require('ws').Server;

var hypervisor = new libvirt.Hypervisor('qemu:///system');

/* Fetch active domains using libvirt. */
var domains;
function updatedomains() {
    var activedomains = hypervisor.getActiveDomains();
    var unsorteddomains = {};
    var domainnames = [];
    for (i=0; i<activedomains.length; i++) {
        domain = hypervisor.lookupDomainById(activedomains[i]);
        var xml = domain.toXml([ libvirt.VIR_DOMAIN_XML_SECURE, libvirt.VIR_DOMAIN_XML_INACTIVE ]);
        /* A dirty hack, but it's all we have for now. */
        var port = xml.replace(/[.\s\S]*<graphics type='spice' port='([0-9]+)' [.\s\S]*/m, '$1');
        unsorteddomains[domain.getName()] = { id: activedomains[i], port: port };
        domainnames.push(domain.getName());
    }
    domainnames.sort();
    domains = {};
    for (i=0; i<domainnames.length; i++)
        domains[domainnames[i]] = unsorteddomains[domainnames[i]];
}

/* Create the li-elements for the html. */
function domainelements(domains) {
    var elements = '';
    var base = '<li id="domain-{name}"><a href="#" onclick="return connect(\'{name}\');">{name}</a></li>';
    for (var name in domains)
        elements = elements+base.replace(/{name}/g, name);
    return elements;
}

var selectprotocol = function(protocols, callback) {
    if (protocols.indexOf('binary')>=0) {
        callback(true, 'binary');
    } else if (protocols.indexOf('base64')>=0) {
        callback(true, 'base64');
    } else {
        console.log('Client must support binary or base64.');
        callback(false);
    }
}

var handleclient = function(client) {
    console.log('Socket client, uri: '+client.upgradeReq.url);
    console.log('Version '+client.protocolVersion+', subprotocol: '+client.protocol);
    if (client.upgradeReq.url.match(/^\/domain\//)) {
        domain = client.upgradeReq.url.replace(/.*\//, '');
        console.log('Connecting to domain '+domain);
        var target = net.createConnection({
            port: domains[domain]['port'],
            host: '127.0.0.1'
        }, function() {
            console.log('Connected to localhost:'+domains[domain]['port']);
        });
        target.on('data', function(data) {
            try {
                if (client.protocol==='base64')
                    client.send(new Buffer(data).toString('base64'));
                else
                    client.send(data, { binary: true });
            } catch (exception) {
                console.log('Server closed connection.');
                target.end();
            }
        });
        target.on('end', function() {
            console.log('Server closed connection.');
            client.close();
        });
        target.on('error', function() {
            console.log('Server error.');
            target.end();
            client.close();
        });
        client.on('message', function(message) {
            if (client.protocol==='base64')
                target.write(new Buffer(message, 'base64'));
            else
                target.write(message, 'binary');
        });
        client.on('close', function(code, reason) {
            console.log('Client closed connection: '+code+' ('+reason+')');
            target.end();
        });
        client.on('error', function(error) {
            console.log('Client error: '+error);
            target.end();
        });
    }
};

var handlerequest = function(request, response) {
    var uri = url.parse(request.url).pathname.replace(/^\/?$/, 'virtconsole.html').replace(/^\//, '');
    var file = path.join('spice-html5', uri);
    if (uri.match(/^virtconsole\.(html|css)$/))
        file = uri;
    fs.exists(file, function(exists) {
        if (!exists) {
            console.log('File not found: '+file);
            response.writeHead(404, { 'Content-Type': 'text/plain' });
            response.write('404 Not Found\n');
            response.end();
            return;
        }
        fs.readFile(file, 'binary', function(error, data) {
            if (error) {
                response.writeHead(500, { 'Content-Type': 'text/plain' });
                response.write(error + '\n');
                response.end();
                return;
            }
            if (file=='virtconsole.html')
                data = data.replace(/{domains}/, domainelements(domains));
            response.writeHead(200);
            response.write(data, 'binary');
            response.end();
        });
    });
};

updatedomains();

var webserver = http.createServer(handlerequest);
webserver.listen(8000, '127.0.0.1', function() {
    socketserver = new WebSocketServer({
        server: webserver,
        handleProtocols: selectprotocol
    });
    socketserver.on('connection', handleclient);
});

File: virtconsole.py
import base64
import libvirt
import os.path
import re
import socket
import tornado.ioloop
import tornado.web
import tornado.websocket


DOMAINS = {}


class HtmlHandler(tornado.web.RequestHandler):

    def get(self, abspath, start=None, end=None):
        with open(os.path.join(os.path.dirname(__file__), 'virtconsole.html'), 'rb') as file:
            content = file.read()
        elements = []
        for name in sorted(DOMAINS.keys()):
            elements.append('<li id="domain-%s"><a href="#" onclick="return connect(\'%s\');">%s</a></li>' % (name, name, name))
        self.write(re.sub(r'{domains}', ''.join(elements), content))


class ProxyServer(object):

    def __init__(self, port, client=None):
        self.port = port
        self.client = client
        self.stream = None
        self.sock_fd = None

    def connect(self):
        self.sock_fd = socket.socket(socket.AF_INET, socket.SOCK_STREAM, 0)
        self.stream = tornado.iostream.IOStream(self.sock_fd)
        self.stream.set_close_callback(self.on_close)
        self.stream.connect(('127.0.0.1', self.port), self.read)

    def on_receive(self, data):
        self.stream.read_bytes(1024, streaming_callback=self.on_streaming, callback=self.on_receive)

    def on_streaming(self, data):
        if self.client.subprotocol=='base64':
            self.client.write_message(base64.b64encode(data))
        else:
            self.client.write_message(data, binary=True)

    def read(self):
        self.stream.read_bytes(1024, streaming_callback=self.on_streaming, callback=self.on_receive)

    def on_close(self):
        if self.client:
            self.client.close()

    def write(self, msg):
        self.stream.write(msg)

    def close(self):
        self.stream.close()


class WebSocketHandler(tornado.websocket.WebSocketHandler):

    clients = set()
    stream_map = {}

    def check_origin(self, domain):
        return True

    def select_subprotocol(self, subprotocols):
        if 'binary' in subprotocols:
            self.subprotocol = 'binary'
        elif 'base64' in subprotocols:
            self.subprotocol = 'base64'
        return self.subprotocol

    def open(self, domain):
        WebSocketHandler.clients.add(self)
        s = socket.socket(socket.AF_INET, socket.SOCK_STREAM, 0)
        c = ProxyServer(DOMAINS[domain]['port'], self)
        self.target = ProxyServer(DOMAINS[domain]['port'], client=self)
        WebSocketHandler.stream_map[self] = c
        c.connect()

    def on_message(self, message):
        st = WebSocketHandler.stream_map[self]
        if self.subprotocol=='base64':
            st.write(base64.b64decode(message))
        else:
            st.write(message)

    def on_close(self):
        WebSocketHandler.clients.remove(self)
        if WebSocketHandler.stream_map.get(self):
            st = WebSocketHandler.stream_map[self]
            st.close()
            del WebSocketHandler.stream_map[self]


if __name__ == '__main__':

    hypervisor = libvirt.openReadOnly('qemu:///system')
    for id in hypervisor.listDomainsID():
        domain = hypervisor.lookupByID(id)
        xml = domain.XMLDesc()
        # A dirty hack, but it's all we have for now.
        port = re.sub(r".*<graphics type='spice' port='([0-9]+)' .*", r'\1', xml, flags=re.DOTALL)
        DOMAINS[domain.name()] = { 'id': id, 'port': int(port) }

    app = tornado.web.Application([
        (r'/(virtconsole\.html)?', HtmlHandler),
        (r'/+domain/(.*)', WebSocketHandler),
        (r'/(.*)', tornado.web.StaticFileHandler, {
            'path': os.path.dirname(__file__),
        }),
    ])
    app.listen(8000)
    tornado.ioloop.IOLoop.current().start()
